### PR TITLE
added `ParentId` to `TelemetrySpan`

### LINKED ISSUE #2733: OpenTelemetry.Api: need to expose `ParentId` somewhere on `TelemetrySpan` or `SpanContext`
# Feature Request

**Is your feature request related to a problem?**
Yes, right now implementing an OTel instrumentation inside a framework using the OpenTelemetry.Api package is difficult because we have no clear way of asserting whether or not parent / child span relationships are preserved correctly due to missing members on the `TelemetrySpan` and `SpanContext` classes.

**Describe the solution you'd like:**

The `Activity.ParentId` or `Activity.ParentSpanId` property should be exposed through a corresponding `TelemetrySpan.ParentId` or `SpanContext.ParentId` member.

**Describe alternatives you've considered.**

Looking into just using the `Activity` API directly, even though my use case is one of the ones where it's recommended that we use the OpenTelemetry.Api shim instead: https://github.com/open-telemetry/opentelemetry-dotnet/tree/6b7f2dd77cf9d37260a853fcc95f7b77e296065d/src/OpenTelemetry.Api#introduction-to-opentelemetry-net-tracing-api

### PR DESCRIPTION
**Pull request by Aaronontheweb**
Fixes #2733

## Changes

Added a `ParentId` property to `TelemetrySpan` in order to make it possible to assert that instrumentations built against OpenTelemetry.Api produce the correct correlations.

For significant contributions please make sure you have completed the following items:

* [x] Appropriate `CHANGELOG.md` updated for non-trivial changes
* [x] Design discussion issue #2733
* [x] Changes in public API reviewed

### CODE REVIEW

**Comment by Aaronontheweb** [comment]
Signed the CLA

File: src/OpenTelemetry.Api/.publicApi/net461/PublicAPI.Shipped.txt [hunk 1]
```diff
@@ -108,6 +108,7 @@ OpenTelemetry.Trace.TelemetrySpan.Dispose() -> void
 OpenTelemetry.Trace.TelemetrySpan.End() -> void
 OpenTelemetry.Trace.TelemetrySpan.End(System.DateTimeOffset endTimestamp) -> void
 OpenTelemetry.Trace.TelemetrySpan.IsRecording.get -> bool
+OpenTelemetry.Trace.TelemetrySpan.ParentId.get -> System.Diagnostics.ActivitySpanId
```
Comment (cijothomas) [inline src/OpenTelemetry.Api/.publicApi/net461/PublicAPI.Shipped.txt:111]: all public changes should go to Unshipped.txt
(It'll be moved to shipped, once we ship a stable version with it)

File: src/OpenTelemetry.Api/Trace/TelemetrySpan.cs [hunk 2]
```diff
@@ -64,6 +64,24 @@ public bool IsRecording
             }
         }
 
+        /// <summary>
+        /// Gets the identity of the parent span id, if any.
+        /// </summary>
+        public ActivitySpanId ParentId
```
Comment (cijothomas) [inline src/OpenTelemetry.Api/Trace/TelemetrySpan.cs:70]: Would it be better to expose parentspancontext instead? (consisting of all fields from SpanContext)

File: src/OpenTelemetry.Api/.publicApi/net461/PublicAPI.Shipped.txt (hunk 1 above)
Comment (Aaronontheweb) [reply src/OpenTelemetry.Api/.publicApi/net461/PublicAPI.Shipped.txt:111]: Ah ok, didn't know

File: src/OpenTelemetry.Api/Trace/TelemetrySpan.cs (hunk 2 above)
Comment (Aaronontheweb) [reply src/OpenTelemetry.Api/Trace/TelemetrySpan.cs:70]: I think so and I tried to do that first, but the `ActivityContext` doesn't include the `ParentId` - that's incorporated onto the `Activity` type itself, so I'd need to change the `SpanContext` constructor to pass that information along as a separate field. I wanted to keep the changes to a bare minimum until I'd spoken with a project member but I'd be happy to re-arrange that.

File: src/OpenTelemetry.Api/CHANGELOG.md [hunk 3]
```diff
@@ -2,6 +2,8 @@
 
 ## Unreleased
 
+* Added `ParentId` to `TelemetrySpan` ([#2733](https://github.com/open-telemetry/opentelemetry-dotnet/issues/2733))_
```
Comment (mic-max) [inline src/OpenTelemetry.Api/CHANGELOG.md:5]:
```suggestion
* Added `ParentId` to `TelemetrySpan` ([#2740](https://github.com/open-telemetry/opentelemetry-dotnet/pull/2740/files))_
```

This should be a link to the pull request, which unfortunately you don't actually get until you make the PR and are given the ID and then have to come back and update the changelog 😅

File: src/OpenTelemetry.Api/Trace/TelemetrySpan.cs (hunk 2 above)
Comment (cijothomas) [reply src/OpenTelemetry.Api/Trace/TelemetrySpan.cs:70]: Yes, SpanContext (ActivityContext) does not include ParentId, and is as per spec, so we don't need to change it.
The spec says Span should have a way to get parent span either in the form of a Span itself, or `SpanContext`.

Activity currently exposes `Parent` (for in-proc parent), and `ParentId` and `ParentSpanId` for out-of-proc parent. The `ParentId` is in the W3CTraceParent format, and we may have to parse it to convert to `Activity/SpanContext`.

If your requirement is met by simply having `ParentSpanId` (as in this PR), lets start with that now, and expose more later. If this is the case, then rename the field to `ParentSpanId` to make it clears its the SpanId only.

Comment (cijothomas) [reply src/OpenTelemetry.Api/Trace/TelemetrySpan.cs:70]:
```suggestion
        public ActivitySpanId ParentSpanId
```

File: src/OpenTelemetry.Api/.publicApi/net461/PublicAPI.Shipped.txt (hunk 1 above)
Comment (Aaronontheweb) [reply src/OpenTelemetry.Api/.publicApi/net461/PublicAPI.Shipped.txt:111]: I'll fix this tonight

Comment (Aaronontheweb) [reply src/OpenTelemetry.Api/.publicApi/net461/PublicAPI.Shipped.txt]: Done

File: src/OpenTelemetry.Api/Trace/TelemetrySpan.cs (hunk 2 above)
Comment (alanwest) [reply src/OpenTelemetry.Api/Trace/TelemetrySpan.cs]: +1 on adding a `ParentSpanContext` to better reflect the spec.

I think you can construct the `SpanContext` for the parent from a `TelemetrySpan` like so:

```csharp
SpanContext? parent = this.Activity.Parent != null
    ? new SpanContext(this.Activity.Parent.Context)
    : null;
```

Comment (Aaronontheweb) [reply src/OpenTelemetry.Api/Trace/TelemetrySpan.cs]: Ok, so is that what we should do? 'ParentContext'?

Comment (cijothomas) [reply src/OpenTelemetry.Api/Trace/TelemetrySpan.cs]: @alanwest `this.Activity.Parent` would be null, if the parent is not an in-proc one. In that case, `Activity.ParentId` could be used to construct the SpanContext, but i am not sure if thats fully correct.

File: src/OpenTelemetry.Api/.publicApi/netstandard2.0/PublicAPI.Shipped.txt [hunk 4]
```diff
@@ -108,6 +108,7 @@ OpenTelemetry.Trace.TelemetrySpan.Dispose() -> void
 OpenTelemetry.Trace.TelemetrySpan.End() -> void
 OpenTelemetry.Trace.TelemetrySpan.End(System.DateTimeOffset endTimestamp) -> void
 OpenTelemetry.Trace.TelemetrySpan.IsRecording.get -> bool
+OpenTelemetry.Trace.TelemetrySpan.ParentId.get -> System.Diagnostics.ActivitySpanId
```
Comment (CodeBlanch) [inline src/OpenTelemetry.Api/.publicApi/netstandard2.0/PublicAPI.Shipped.txt:111]: This change should go into the "Unshipped" file, not the "Shipped" file.

Comment (Aaronontheweb) [reply src/OpenTelemetry.Api/.publicApi/netstandard2.0/PublicAPI.Shipped.txt]: Thought I'd removed it. Done.

File: src/OpenTelemetry.Api/Trace/TelemetrySpan.cs (hunk 2 above)
Comment (alanwest) [reply src/OpenTelemetry.Api/Trace/TelemetrySpan.cs]: Ah yes that's right. Sorry 😅 you said this, but I didn't quite follow at first, thanks for the clarification. Agreed, makes sense to move forward with a `ParentSpanId` property.

File: src/OpenTelemetry.Api/CHANGELOG.md [hunk 5]
```diff
@@ -2,6 +2,8 @@
 
 ## Unreleased
 
+* Added `ParentId` to `TelemetrySpan` ([#2740](https://github.com/open-telemetry/opentelemetry-dotnet/pull/2740))
```
Comment (alanwest) [inline src/OpenTelemetry.Api/CHANGELOG.md:5]:
```suggestion
* Added `ParentSpanId` to `TelemetrySpan` ([#2740](https://github.com/open-telemetry/opentelemetry-dotnet/pull/2740))
```

Comment (Aaronontheweb) [reply src/OpenTelemetry.Api/CHANGELOG.md]: Whoops, missed that spot. Added your suggestion.

**Review by alanwest — APPROVED** [review]
LGTM

**Comment by Aaronontheweb** [comment]
Did I do the API approval wrong or something? Looks like that's what the test suite is complaining about.

**Comment by cijothomas** [comment]
> Did I do the API approval wrong or something? Looks like that's what the test suite is complaining about.

Public API file for only one target is updated, thats likely why CI is failing. There should be similar changes for each target frameworks.

**Comment by Aaronontheweb** [comment]
> > Did I do the API approval wrong or something? Looks like that's what the test suite is complaining about.
> 
> Public API file for only one target is updated, thats likely why CI is failing. There should be similar changes for each target frameworks.

Fixed

**Comment by Aaronontheweb** [comment]
Looks like the test failures here are unrelated to my changes. What do we do?

File: test/OpenTelemetry.Tests/Trace/TelemetrySpanTest.cs [hunk 6]
```diff
@@ -67,5 +67,22 @@ public void CheckRecordExceptionEmpty()
             telemetrySpan.RecordException(null);
             Assert.Empty(activity.Events);
         }
+
+        [Fact]
+        public void ParentIds()
+        {
+            using var activitySource = new ActivitySource("testSource");
+            using var parentActivity = activitySource.CreateActivity("parentOperation", ActivityKind.Internal);
```
Comment (cijothomas) [inline test/OpenTelemetry.Tests/Trace/TelemetrySpanTest.cs:75]: without any listeners, this activity will not be created, so thats causing the unit test failures.

Comment (Aaronontheweb) [reply test/OpenTelemetry.Tests/Trace/TelemetrySpanTest.cs:75]: Thanks! I'll take a look

Comment (Aaronontheweb) [reply test/OpenTelemetry.Tests/Trace/TelemetrySpanTest.cs]: I believe I fixed this - didn't get any `dotnet test` failures back when running locally.

**Comment by Aaronontheweb** [comment]
Thank you!

**Pull request merged**